### PR TITLE
Update migrations.rake

### PR DESCRIPTION
**Pull request by inner-whisper**
Fix error for multiple hosts deploy (described here https://github.com/capistrano/capistrano/issues/1449).
Now `deploy:migrate` work with specific host (`cap production HOSTS=example.com deploy:migrate`  or  `cap production HOSTS=example1.com deploy:migrate` ) as well as with all hosts (`cap production deploy:migrate`)

### CODE REVIEW

**Comment by mattbrictson** [comment]
I understand that yours is a special case, but normally Rails deployments should never run migrations on multiple servers. Your change will therefore break the default behavior for a lot of users. Is there a way you can make your change opt-in?

If you make it opt-in I think your PR will have a better chance of being accepted.

For example, you could introduce an intermediate setting that defaults to `primary fetch(:migration_role)`, but that could be overridden for edge cases like yours.

**Comment by tonkonogov** [comment]
Exactly what I did a month ago on my fork there https://goo.gl/SqMY1p. Hope @inner-whisper will take my changes for this PR.

**Comment by inner-whisper** [comment]
@tonkonogov, I think that your changes was suggested in that PR - https://github.com/capistrano/rails/pull/126

@mattbrictson, can you explain use case of current setting `on primary fetch(:migration_role)`?
With this setting we have the limitation: use role `db` (or another migration role) only with one host for migration. With setting `on roles(fetch(:migration_role))` we can use limited list of hosts, where capistrano will try to migrate db (`role :db,  'host1', 'host2', 'host3'` will migrate all listed hosts). Why we need to limit this list only with first host in `db` role hosts list (`role :db,  'host1', 'host2', 'host3'` will migrate only 'host1')?

**Comment by tonkonogov** [comment]
You need one host to run because in general all hosts from the environment have the only database, therefore you need to run migration only once in order to alter a database. 

My changes are based on #126, but I've added conditional roles list as well. It doesn't change default behaviour described by @mattbrictson .

**Comment by mattbrictson** [comment]
> @mattbrictson, can you explain use case of current setting `on primary fetch(:migration_role)`?

@inner-whisper The typical reasons Rails apps are deployed on multiple servers is for horizontal scale. Usually this means one database and many app servers with a load balancer in front. Each app server runs an identical copy of the app, and shares the single DB. This is the scenario Capistrano was originally designed for.

In that scenario, you only want to run migrations once for the entire cluster. Running the migrations in parallel on multiple machines would result in errors, or worse.

You might also have a primary and secondary database, with some sort of replication strategy. In that case, you still only want to run migrations once (i.e. on the primary db), because the replication takes care of propagating the changes to the secondary db.

In other words, `on primary fetch(:migration_role)` is the desired behavior for all these scenarios.

**Comment by carlesso** [comment]
Just my 2 cents. I have worked on different project where I had to face both scenarios.

### Multiple app one db
It comes natural to me to set `db` role only on one of the server

### Multiple instances of app (separate dbs)
I used to have multiple stage file for each instance, but it is not possible from most continuous delivery platform to deploy the same branch with different capistrano stages.

So I've created just one `production.rb` file having 3 servers all with `db` role, expecting the migrations to run on all machines. That is what comes intuitive to me. I had to dig the internet to find out what was going wrong, also because calling `HOSTS` filter and `deploy:migrate` directly had no effect. 

---

All this just to say that sometimes you need to have a single capistrano stage for multiple instances, and right now it's not possible.

**Comment by will-in-wi** [comment]
I think this PR is superseeded by https://github.com/capistrano/rails/pull/168 which was merged.

Close?
